### DISCUSSION TITLE
Flatbuffer captures from rooted LG TV

### DISCUSSION THREAD

**Discussion by tavalin** [awawa-dev/HyperHDR#169] [General]
Hi @awawa-dev - I have managed to root my LG TV to natively capture all displayed content and send to HyperHDR using Flatbuffers using [https://github.com/TBSniller/piccap](url)

I have a couple of questions:

- Is it possible to apply the same image processing to screen captures received through Flatbuffers as captured through a USB grabber?
- Is it possible to show the fps for screen captures received through Flatbuffers?

Thanks

**Comment by awawa-dev** [comment]
Hi
> Is it possible to apply the same image processing to screen captures received through Flatbuffers as captured through a USB grabber?

Currently no, only direct processing images from flatbuffer/protobuffer frames to the LED strip. And what would be purpose of that? Piccap isn't supposed to provide already processed & decoded frames from HDR or DV stream? Or it's a raw video content without any HDMI meta-data applied just like in case of USB grabbers?

> Is it possible to show the fps for screen captures received through Flatbuffers?

Not directly, but current v18beta on github has performance statistics for rendered frames by the LED drivers in logs. So when you disable the smoothing it should show the result: flatbuffer rate including time needed to process the images and to send LED data to devices. More performance counters will be provided in the incoming commit, but it will take some time to finish & test it before releasing it for public.

**Reply by tavalin** [reply]
> Piccap isn't supposed to provide already processed & decoded frames from HDR or DV stream? Or it's a raw video content without any HDMI meta-data applied just like in case of USB grabbers?

I did some more testing by altering the picture mode on the TV and it alters screen capture colours etc so it looks like it captures the post-processing image so maybe there isn't much point in further applying further HyperHDR image processing.  I haven't tried it with HDR/DV content yet to see what HDR content looks like.

> So when you disable the smoothing it should show the result: flatbuffer rate including time needed to process the images and to send LED data to devices

I'll try this as the FPS from piccap seems quite low at the moment but I want to verify this first

**Comment by chbartsch** [comment]
Hi awawa, I'm also using Piccap and HDR/DV images look dim/pale in certain situations. With Netflix on a Nvidia Shield for example whereas playing HDR content via Emby on the same device (Shield) looks way better.

Have you already done any work on applying tone mapping to flatbuffer images? Because if not I will try to implement it and see if things get better. I think a simple tone curve adjustment could already be enough in that case.

Already dug through your code a little bit and found `ImageResampler::applyLUT()`. Running this on the flatbuffer images (with a custom LUT?!) would do it, right?

**Reply by awawa-dev** [reply]
Hi
It won't work without loading 50MB LUT table into memory first. That dynamic mechanism is handled by the video & system grabber internally and their LUT table can't be shared with the flatbuffer module.

**Reply by chbartsch** [reply]
Thanks for your very fast reply!! I already made some progress: Applying tone mapping on flatbuffers data works. :) But currently it's always on. So next step will be to apply the global HDR switch and then (best case scenario) have the TV switch HDR-Mode in HyperHDR on/off depending on the content played. I know I can use the JSON API for example but still have to figure out how to get the information about the current data format from the TV. 

Thanks for all the work you put into HyperHDR!!

**Reply by awawa-dev** [reply]
Good job. I took a look on your code. I think some things need to check first: does flatbuffer really need HDR tone mapping or some other transformation? OK, the image look dim but why and how to fix it? For USB grabber it's simply: missed BT2020 transformation and hybrid gamma. That must be explained first, because DV images provided by piccap also are dimmed but why Dolby Vision needs HDR mapping? It's completely different technology. We can provide only one special RGB LUT table globally for all grabbers and by default it does HDR tone mapping or for color correction for grabebrs like special ezcap320/av access. Two other LUT tables are for YUV and HDR tone mapped YUV. 
There is a need to provide code that loads LUT table for macOS and Windows also. I dont want that enabling HDR tone mapping globally (that is on default) would cause automatic loading LUT table for flatbuffer (also enabled on default) for users who don't need it. So probably special option.
 And finally: after testing framebuffer network performance have really bad opinion about that solution which is causing heavy network saturation. Have some idea how to replace it.

**Reply by awawa-dev** [reply]
Did you try to test processing settings like Saturation, Luminance (available when Classic HyperHDR calibration is enabled)  and gamma on the live stream piccap when TV and LED strip were on? Calibration settings are poorly visible in the HyperHDR video preview and must be live tested. They don't improve the effect in an acceptable way?

**Reply by chbartsch** [reply]
> I think some things need to check first: does flatbuffer really need HDR tone mapping or some other transformation? OK, the image look dim but why and how to fix it? For USB grabber it's simply: missed BT2020 transformation and hybrid gamma. That must be explained first, because DV images provided by piccap also are dimmed but why Dolby Vision needs HDR mapping? It's completely different technology.

Honestly I'm not sure about this. Possibly we do not need full HDR tone mapping. For my testing I generated a LUT with Hable (HDR) and all HDR gamma values at 2.2 (everything else default) and this already gave me good results. Maybe this could be achieved in a simpler way? I also thought about applying some post processing to the images directly on the TV but couldn't find any suitable library to do this. And since I'm not much into image processing I thought using something which is already there (your tone mapping stuff) would be easier.

> There is a need to provide code that loads LUT table for macOS and Windows also.

I don't understand this part. How is loading the LUT different on these systems? Shouldn't QFile already manage this?

> I dont want that enabling HDR tone mapping globally (that is on default) would cause automatic loading LUT table for flatbuffer (also enabled on default) for users who don't need it. So probably special option.

Ahh... Just found out that HDR tone mapping will be reenabled on each restart. OK, that is a problem indeed.

> And finally: after testing framebuffer network performance have really bad opinion about that solution which is causing heavy network saturation. Have some idea how to replace it.

Yes, unfortunately network bandwidth is high with flatbuffers. :/ Do you have a better idea for the data exchange between Piccap and HyperHDR? Keep in mind that we are somewhat limited in processing power on the TVs. Is there any fast compression or something that we could use?

**Reply by chbartsch** [reply]
> Did you try to test processing settings like Saturation, Luminance (available when Classic HyperHDR calibration is enabled) and gamma on the live stream piccap when TV and LED strip were on? Calibration settings are poorly visible in the HyperHDR video preview and must be live tested. They don't improve the effect in an acceptable way?

To be honest I must have overlooked saturation and luminance settings so far. Will test how these work tonight.

But as I said my testing LUT was looking good already and also colors did fit my subjective impression. So if the LUT works you still think this could also be achieved with saturation/luminace?

**Reply by awawa-dev** [reply]
> I don't understand this part. How is loading the LUT different on these systems? Shouldn't QFile already manage this?

The lut table could be in different location. Fallback /usr/share/hyperhdr/lut/lut_lin_tables.3d does not apply for the other systems, probably some "#ifdef" will be needed. For example: in Windows in the installation folder on macOS it be even accessible from DMG container. Take a look on [MFGrabber](https://github.com/awawa-dev/HyperHDR/blob/91f12c1febaca2e7414b4efa1d72e4ff24562a62/libsrc/grabber/MF/MFGrabber.cpp#L133) (windows, even similar but no /usr/ fallback is possible) and [AVFGrabber](https://github.com/awawa-dev/HyperHDR/blob/91f12c1febaca2e7414b4efa1d72e4ff24562a62/libsrc/grabber/AVF/AVFGrabber.mm#L146) (macOS, that's quite different).

> Yes, unfortunately network bandwidth is high with flatbuffers. :/ Do you have a better idea for the data exchange between Piccap and HyperHDR? Keep in mind that we are somewhat limited in processing power on the TVs. Is there any fast compression or something that we could use?

Yes I have an idea for smaller frame. And to migrate to UDP which is a proper protocol for network video stream. Almost finished and will be provided in v18beta2 soon if it passed the final tests.

> But as I said my testing LUT was looking good already and also colors did fit my subjective impression. So if the LUT works you still think this could also be achieved with saturation/luminace?

I don't use piccap so can only guess. Really don't know what is LG internal capturing library doing to these dim images, especially DV is a riddle. Saturation/luminace/gamma is a solution that you can already test. Not sure about the result (didn't work well for HDR captured by USB grabbers) but I think it's worth to check it out. Some dynamic adjustment of saturation/luminance/gamma depending on the video source (SDR/HDR/DV) would still be needed but it also can be achieved using JSON API: you can test it and build command with help of HyperHDR JSON playground tab.

**Reply by chbartsch** [reply]
> The lut table could be in different location. Fallback /usr/share/hyperhdr/lut/lut_lin_tables.3d does not apply for the other systems, probably some "#ifdef" will be needed. For example: in Windows in the installation folder on macOS it be even accessible from DMG container. Take a look on [MFGrabber](https://github.com/awawa-dev/HyperHDR/blob/91f12c1febaca2e7414b4efa1d72e4ff24562a62/libsrc/grabber/MF/MFGrabber.cpp#L133) (windows, even similar but no /usr/ fallback is possible) and [AVFGrabber](https://github.com/awawa-dev/HyperHDR/blob/91f12c1febaca2e7414b4efa1d72e4ff24562a62/libsrc/grabber/AVF/AVFGrabber.mm#L146) (macOS, that's quite different).

I see. Just commited a fix. Can't test on Windows or Mac right now but think it should work.

> Yes I have an idea for smaller frame. And to migrate to UDP which is a proper protocol for network video stream. Almost finished and will be provided in v18beta2 soon if it passed the final tests.

Nice. Looking forward to it. Have you developed this protocol yourself or are there some sources available somewhere so that I could start thinking about the sender side implementation?!

> I don't use piccap so can only guess. Really don't know what is LG internal capturing library doing to these dim images, especially DV is a riddle. Saturation/luminace/gamma is a solution that you can already test. Not sure about the result (didn't work well for HDR captured by USB grabbers) but I think it's worth to check it out. Some dynamic adjustment of saturation/luminance/gamma depending on the video source (SDR/HDR/DV) would still be needed but it also can be achieved using JSON API: you can test it and build command with help of HyperHDR JSON playground tab.

I did some testing with saturation/luminance and this also gave me good results. I'm honestly unsure if I like this better than tone mapping or not. Have to do some more testing and write a API based test-script which let me switch fast between saturation/luminance and tone mapping approach to have a proper A/B comparision.

**Reply by awawa-dev** [reply]
> I see. Just commited a fix. Can't test on Windows or Mac right now but think it should work.

I can test it on mac and windows. It's only a question of finding the LUT table (built-in or users) in the correct places. The logic is shared across all platforms and it should be fine.

> Nice. Looking forward to it. Have you developed this protocol yourself or are there some sources available somewhere so that I could start thinking about the sender side implementation?!

The mechanism will include both sender and receiver. And documented protocol. Still need live testing and fine tuning across all platforms and UDP can be chimeric especially when it comes to frame size.

> I did some testing with saturation/luminance and this also gave me good results. I'm honestly unsure if I like this better than tone mapping or not. Have to do some more testing and write a API based test-script which let me switch fast between saturation/luminance and tone mapping approach to have a proper A/B comparision.

Consider also changing values of gamma. Values below 1 should also bump brightness. I know that the calibration process is time-consuming, requires many attempts and the effect is difficult to assess.

**Comment by awawa-dev** [comment]
> Ahh... Just found out that HDR tone mapping will be reenabled on each restart. OK, that is a problem indeed.

Don't bother with it, I just notice it. I can implement the interface configuration for flatbuffer HDR on/off later. It's even better if you focus only on include/flatbufserver and libsrc/flatbufserver (only the LUT file path location depending on the system is missing for now), because I will have to make force push later and some files will change location. Or you can commit PR now and we will adjust it for future changes.

**Reply by chbartsch** [reply]
> Don't bother with it, I just notice it. I can implement the interface configuration for flatbuffer HDR on/off later. It's even better if you focus only on include/flatbufserver and libsrc/flatbufserver (only the LUT file path location depending on the system is missing for now), because I will have to make force push later and some files will change location. Or you can commit PR now and we will adjust it for future changes.

Damn, I missed this message yesterday. I already went ahead and implemented the tone mapping switch into flatbuffers configuration page.

What should I do now? Submit a PR as is or should I remove the switch from flatbuffers configuration first?

**Reply by awawa-dev** [reply]
No, you can publish PR now. When the master branch is changed then github should adjust PR automatically. Automatic merge almost for sure will be not possible but don't worry about that. I will fix it.

**Comment by awawa-dev** [comment]
I've already seen similar distortion. Anyway new LUT table (resources\lut\for_webos_hdr_flat_lut_lin_tables.tar.xz) fixes it in some extent. Must be installed manually in home folder:  
![lut_final](https://user-images.githubusercontent.com/69086569/160293770-7bec577b-4366-4164-96b3-de0d4c85fbf3.png)
LG CX 9 HDR10: before and after. 1080p HDR and 4k HDR resolution produce almost identical result.

**Reply by pho084** [reply]
hi awawa

could you share the settings that you used for the flatbuffers LUT? 
it works great for my android Chromecast in DV/HDR but when i connect my Windows PC in HDR the picture is still too dark and colors could be more vibrant.

**Reply by awawa-dev** [reply]
Strange, it was designed for Windows PC 1080p 10bit RGB (or 12bit don't remember, 23.97Hz) HDR mode. Maybe you automatically activated some special mode like Game (for PC) or different settings like HDR mode in the TV menu (it can be set separately even for each HDMI port)? What's happening inside TV is a mystery for us and even the backend was created by reverse engineering. Anyway I'm working on completely new and universal approach: don't know the final result yet and it won't help if there is a need of using 3 or 4 different LUT tables for each specific situation ;)

**Reply by pho084** [reply]
thank you for your reply. It's weird indeed, but with Windows and HDR i have learned that all bets are off, HDR mode in Windows is just another horrible implementation by Microsoft... anyway, i managed to solve my problem by creating a new LUT and increasing the RGB GAMMA values resulting in a brighter image. i am stoked to see what you have planned for the universal LUT approach though :)
